### PR TITLE
[EDX-162] Partial definition list interaction bugfix

### PR DESCRIPTION
**Pull request by MymmiJ**
> _NOTE TO REVIEWERS_ - please **do not review PRs in the `DRAFT` state**, as the PR may change substantially before it is ready to review. Thanks.

## Description

A PR description indicating the purpose of the PR.

* [Jira ticket](https://ably.atlassian.net/browse/EDX-162).

**Minor/Incidental Changes:**
* Refactor: Move Gatsby/node files to TypeScript
* Refactor: Move html-parser files to TypeScript
* Fix: Newlines inside series of dl elements are compressed into one newline

## Review

Ensure there's no gaps between definition lists on this page:

* [Page to review](http://localhost:8000/docs/api/realtime-sdk)

What you should _not_ see is the gap between 'tokenDetails' and 'tls':

Screenshot 2022-06-22 at 14 03 14

### CODE REVIEW

File: data/html-parser/index.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+type Attributes = { [attr: string]: string };
+
+type HtmlDataType = keyof typeof HtmlDataTypes;
+
+/**
+ * Data that we need to check from any cheerio element to see what to do with it
+ */
+type AnyCheerioElement = {
+  type?: null | undefined | 'text' | 'tag' | 'script' | 'style' | 'comment';
```
Comment (DuncanEdwards) [inline data/html-parser/index.ts:18]: Don't need question mark + undefined, remove undefined from this list (but not null)

File: data/html-parser/index.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  data?: undefined | string;
+  attribs?: null | undefined | Attributes;
+  children: cheerio.Element[];
+};
+
+type ParsedNode = {
+  data: string | ParsedNode[];
+  type: typeof HtmlDataTypes.text | typeof HtmlDataTypes.tag;
+  name: string;
+  attribs?: null | undefined | Attributes;
```
Comment (DuncanEdwards) [inline data/html-parser/index.ts:29]: As above

File: data/html-parser/index.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+type HtmlDataType = keyof typeof HtmlDataTypes;
+
+/**
+ * Data that we need to check from any cheerio element to see what to do with it
+ */
+type AnyCheerioElement = {
+  type?: null | undefined | 'text' | 'tag' | 'script' | 'style' | 'comment';
+  name: HtmlDataType;
+  data?: undefined | string;
```
Comment (DuncanEdwards) [inline data/html-parser/index.ts:20]: As anove

File: data/html-parser/index.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  type: typeof HtmlDataTypes.text | typeof HtmlDataTypes.tag;
+  name: string;
+  attribs?: null | undefined | Attributes;
+};
+
+export const cheerioNodeParser = (_i: number, node: cheerio.Element): ParsedNode => {
+  const {
+    type = null,
+    name = <HtmlDataType>HtmlDataTypes.text,
+    data = '',
```
Comment (DuncanEdwards) [inline data/html-parser/index.ts:36]: If the default is a value, is this data? or is the default value undefined?

**Review by DuncanEdwards — CHANGES_REQUESTED** [review]
Some minor frittering from Duncan

File: data/html-parser/remove-paragraphs-from-definition-lists.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+export const removeParagraphsFromDefinitionListsAndMerge = (cheerioNodes: cheerio.Selector) =>
+  cheerioNodes('dl').each((_i: number, elem: cheerio.Element) => {
+    // As jQuery/Cheerio is stateful and imperative, including comments to clarify
+    // Remove empty paragraphs immediately after any dl
+    const siblings = cheerioNodes(elem)
+      .siblings('p')
+      .filter((_i, elem) => {
```
[the remark this answers is not in the document]
Comment (MymmiJ) [reply data/html-parser/remove-paragraphs-from-definition-lists.ts:7]: Indicating that the index variable isn't used, it's a pattern  in the EsLint config we've got in docs for unused variables (but can easily change if needed)

File: data/html-parser/index.ts (hunk 1 above)
Comment (MymmiJ) [reply data/html-parser/index.ts]: Cool! Removed.

File: data/html-parser/index.ts (hunk 4 above)
Comment (MymmiJ) [reply data/html-parser/index.ts]: You're right, I was flailing trying to get Types to compile properly, I think this was an artefact of a previous attempt; have removed redundant defaults and all works fine.

File: data/html-parser/remove-paragraphs-from-definition-lists.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (DuncanEdwards) [reply data/html-parser/remove-paragraphs-from-definition-lists.ts:7]: Of course! Yeah, I changed them to just underscore.

**Review by DuncanEdwards — APPROVED** [review]

**Pull request merged**